File: 6-zigzag-conversion/zigzag-conversion.swift
class Solution {
    func convert(_ s: String, _ numRows: Int) -> String {
        if numRows == 1 { return s }
        if numRows >= s.count { return s }
        
        var rows: [[Character]] = []
        for i in 0 ..< numRows {
            rows.append([])
        }
        
        var currentRowIndex = 0
        var step = 1
        
        for char in s {
            rows[currentRowIndex].append(char)
            currentRowIndex += step
            
            if currentRowIndex == 0 || currentRowIndex == numRows - 1 {
                step = -step
            }
        }
        
        var result = ""
        for i in 0 ..< rows.count {
            let row = rows[i]
            result.append(.init(row))
        }
        
        return result
    }
}

File: 6-zigzag-conversion/zigzag-conversion.rs
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        if num_rows <= 1 || num_rows >= s.len() as i32{
            return s;
        }
        let chars: Vec<char> = s.chars().collect();
        let mut result:String = String::new();
        let diag_size: i32 = num_rows - 2;
        let step: i32 = num_rows + diag_size;
        for i in 0..num_rows{
            let k: i32 = if i == 0 || i == num_rows - 1 { 0 } else { i * 2 };
            result.push(chars[i as usize]);
            let mut j = i + step;
            while j < s.len() as i32{
                if k != 0 && j - k != i && i < num_rows - 1{
                    result.push(chars[(j - k) as usize])
                }
                result.push(chars[j as usize]);
                j += step
            }
            if i < num_rows - 1 && j >= s.len() as i32 && j - k < s.len() as i32{
                result.push(chars[(j - k) as usize])
            }
        }
        result
    }
}